golang/cosmos/app/app.go

```diff
@@ -803,14 +804,88 @@ func NewAgoricApp(
+	return retrieveSwingStoreExport()
+}
+
+func (eventHandler swingStoreMigrationEventHandler) OnExportRetrieved(provider swingsetkeeper.SwingStoreExportProvider) (err error) {
+	defer helpers.RecoverToError(&err)
+
+	var exportDataIterator sdk.Iterator
+	exportDataIterator, err = provider.GetExportDataIterator()
+	if err != nil {
+		return err
+	}
+	defer exportDataIterator.Close()
+
+	if !exportDataIterator.Valid() {
+		return fmt.Errorf("swing store export must provide some export data")
+	}
+
+	for ; exportDataIterator.Valid(); exportDataIterator.Next() {
+		eventHandler.swingStore.Set(exportDataIterator.Key(), exportDataIterator.Value())
+	}
+	return nil
+}
+
 // upgrade11Handler performs standard upgrade actions plus custom actions for upgrade-11.
 func upgrade11Handler(app *GaiaApp, targetUpgrade string) func(sdk.Context, upgradetypes.Plan, module.VersionMap) (module.VersionMap, error) {
 	return func(ctx sdk.Context, plan upgradetypes.Plan, fromVm module.VersionMap) (module.VersionMap, error) {
 		app.CheckControllerInited(false)
 		// Record the plan to send to SwingSet
 		app.upgradePlan = &plan
 
-		// TODO: Migrate x/vstorage swingStore to x/swingset SwingStore
+		savedBlockHeight := uint64(ctx.BlockHeight() - 1)
+
+		// First, repair swing-store metadata in case this node was state-synced
+		// This is done without checks on the block height as it's idempotent,
+		// and only missing historical metadata is imported
+		getExportDataIterator := func() (iter sdk.Iterator, err error) {
+			defer helpers.RecoverToError(&err)
+
+			return swingsetkeeper.NewExportDataIterator(swingsetkeeper.NewVstorageDataEntriesReader(
+				app.VstorageKeeper.ExportStorageFromPrefix(ctx, swingsetkeeper.StoragePathSwingStore),
+			)), nil
+		}
+
+		readArtifact := func() (artifact swingsettypes.SwingStoreArtifact, err error) {
+			return artifact, io.EOF
+		}
+
+		err := app.SwingStoreExportsHandler.RestoreExport(
+			swingsetkeeper.SwingStoreExportProvider{
+				BlockHeight:           savedBlockHeight,
+				GetExportDataIterator: getExportDataIterator,
+				ReadArtifact:          readArtifact,
+			},
+			swingsetkeeper.SwingStoreRestoreOptions{
+				ExportDataMode: swingsetkeeper.SwingStoreRepairMetadataExportDataMode,
+				ArtifactLevel:  swingsetkeeper.SwingStoreNoneArtifactLevel,
+			},
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		// Then migrate the swing-store shadow copy
+		app.VstorageKeeper.RemoveEntriesWithPrefix(ctx, swingsetkeeper.StoragePathSwingStore)
+		err = app.SwingStoreExportsHandler.InitiateExport(
+			savedBlockHeight,
+			swingStoreMigrationEventHandler{swingStore: app.SwingSetKeeper.GetSwingStore(ctx)},
+			swingsetkeeper.SwingStoreExportOptions{ArtifactLevel: "none", ExportDataMode: "all"},
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		err = swingsetkeeper.WaitUntilSwingStoreExportDone()
+		if err != nil {
+			return nil, err
+		}
```

Comment:
This comment could use some elaboration, e.g.
```suggestion
		// Then migrate the swing-store shadow copy:
		// 1. Remove swingStore-shadowing entries from vstorage.
		// 2. Export swingStore data (as of the previous block) through a handler
		//    that writes the key-value pairs into the swingset module.
		app.VstorageKeeper.RemoveEntriesWithPrefix(ctx, swingsetkeeper.StoragePathSwingStore)
		err = app.SwingStoreExportsHandler.InitiateExport(
			savedBlockHeight,
			swingStoreMigrationEventHandler{swingStore: app.SwingSetKeeper.GetSwingStore(ctx)},
			swingsetkeeper.SwingStoreExportOptions{ArtifactLevel: "none", ExportDataMode: "all"},
		)
		if err == nil {
			err = swingsetkeeper.WaitUntilSwingStoreExportDone()
		}
		if err != nil {
			return nil, err
		}
```

Reply:
Let me know if the new comments are more clear